Working on your own & show-and-tell chapters (#67)
jupytercon/jupytercon2025-developingextensions: 04-materials/06-working-on-your-own.md

All paths are valuable! Pick what excites you most.
:::

## 🚀 Getting Started

Before diving into your chosen path, make sure you have:

- A clear goal (extension idea or GitHub issue to work on)
- Development environment ready (if building new: follow [Setup](02-anatomy-of-extensions.md#️-setup) and [Exercise A](02-anatomy-of-extensions.md#️-exercise-a-15-minutes-extension-creation-and-development-loop))
- Access to [JupyterLab API docs](https://jupyterlab.readthedocs.io/en/latest/api/) and [Extension Examples](https://github.com/jupyterlab/extension-examples)

## 🔨 Path 1: Build your own extension from scratch

Create something entirely new using the extension template, AI assistance, and the patterns you've learned.

### ⚡ Quick start

Follow these sections from chapter 2 to start a new extension from a template:

1. [🛠️ Setup](02-anatomy-of-extensions.md#️-setup) - Set up your environment, Git, and GitHub repository
2. [🏋️ Exercise A: Extension creation and development loop](02-anatomy-of-extensions.md#️-exercise-a-15-minutes-extension-creation-and-development-loop) - Create your extension from the template and test the development loop

> Once you have your extension set up and working, return here to choose an inspiration project below.

Reviewer: ```suggestion
Once you have your extension set up and working, proceed to work on your extension idea or return here to find inspiration below.
```

### 💫 Inspiration

The following three examples with prompts have been tested and confirmed working. Pick one that matches your interests - whether you want visual effects, creative customization, or full-stack development. Use them as-is to practice prompt engineering, or as inspiration for your own ideas.

#### 🎉 Confetti celebration button

**Complexity**: Beginner (Frontend only)
**Time**: 15-20 minutes

> **What you'll learn**: Status bar integration, DOM manipulation, visual effects

Reviewer 2: ```suggestion
**What you'll work with**: Status bar integration, DOM manipulation, visual effects
```

**The Prompt**:
```
Add a button in the status bar to celebrate something! It should show confetti
on top of the UI for 3 seconds
```

**Expected Results**:
- Status bar shows a celebration button (usually bottom-right)
- Clicking triggers confetti animation overlay
- Animation appears on top of all UI elements
- Confetti clears after ~3 seconds
- Multiple clicks work correctly

**What to Watch For**:

> If the confetti doesn't appear:

Reviewer: ```suggestion
If the confetti doesn't appear, try prompting:
```

```
The button is there but I don't see confetti. Check that the z-index
is high enough and the confetti container is properly positioned.
```

To customize:
```
Make the confetti more colorful and add a sound effect when it triggers.
```

**Verified Setup**:
- ✅ macOS 15.7, Claude Code (Claude Sonnet 4.5)
- ✅ `conda` environment: Python 3.13, Node.js 22, JupyterLab
- 📹 [Watch the demo](https://www.loom.com/share/2afabea0184045fa868271f9ab0ca083)

:::{tip}
This is a perfect first project - it's self-contained, purely visual, and you'll immediately see if it works. It teaches you about JupyterLab's status bar API without backend complexity.
:::

#### 🎨 Custom theme extension

**Complexity**: Beginner (Frontend only)
**Time**: 20-30 minutes
**What you'll learn**: Theme customization, CSS styling, JupyterLab theming system

**The Prompt**:
```
Create a theme based on [your favorite movie/show/game/aesthetic]
```

**Example**:
```
Create a theme based on Netflix show KPop Demon Hunters
```

**Expected results**:
- A new theme appears in Settings → Theme menu
- Theme includes custom colors matching your chosen aesthetic
- JupyterLab interface reflects the theme's visual style

> - Theme can be toggled on/off

Reviewer: ```suggestion
- Theme can be toggled on/off in Settings → Theme menu
```

Or we can omit this since bullet 1 already mentions the theme being selectable in this menu.

https://www.billboard.com/wp-content/uploads/2025/07/kpop-demon-hunters-billboard-1800.jpg?w=942&h=628&crop=1
```

To refine the styling:
```
The background image is too bright - make it more subtle with reduced opacity
and add a dark overlay so text is readable.
```

To adjust colors:
```
Update the sidebar and toolbar colors to match the theme's color palette.
Use [specific colors] for accent elements.
```

**Tips for finding images**:
- Search "[theme name] wallpaper 4k" for high-quality backgrounds
- Look for images with good contrast for text readability
- Consider color palettes - extract 3-5 main colors from your chosen image (AI can help you with it!)
- Free image sources: Unsplash, Pexels, Wallhaven

:::{tip}
This is a perfect creative project! You get immediate visual feedback, can personalize your JupyterLab environment, and learn how JupyterLab's theming system works. Plus, you'll have a custom theme you actually want to use daily.

Themes are also great conversation starters - share your theme with other workshop participants!
:::

#### 📊 CPU monitor widget

**Complexity**: Intermediate (Frontend + Server)
**Time**: 30-40 minutes

> **What you'll learn**: REST API integration, backend data processing, graceful error handling

Reviewer 2: ```suggestion
**What you'll work on**: REST API integration, backend data processing, error handling
```

If we are moving away from learning objectives, maybe we can phrase this slightly differently?

- ✅ macOS 15.7, Cursor (Claude Sonnet 4.5 MAX)
- ✅ `conda` environment: Python 3.13, Node.js 22, JupyterLab
- ⚠️ Temperature data gracefully handled as N/A on macOS (expected)
- 📹 [Watch the demo](https://www.loom.com/share/9f6d11d537a94a30af7559fd4d80eea2)

:::{tip}
This teaches you the full stack: backend API design, frontend-backend communication, error handling, and periodic updates. It's a perfect template for any monitoring or dashboard extension.
:::

### 🌟 More extension ideas

Pick an idea that matches your comfort level and interests:

#### 🎓 Beginner (frontend only)
1. **Theme switcher dropdown**: Add a quick theme selector to the toolbar for easy switching
2. **Quote of the day**: Main area widget that fetches and displays random quotes
3. **Keyboard shortcut viewer**: Panel showing all available shortcuts
4. **Pomodoro timer**: Status bar timer for focused work sessions with notifications

#### 🚦 Intermediate (frontend + server)
1. **File size analyzer**: Scan workspace directory and show largest files/folders
2. **Git status widget**: Display current branch, uncommitted changes count
3. **Environment inspector**: Show installed packages and Python version

> 4. **Todo list with persistence**: Sidebar panel that saves tasks to disk

Author: I thought disk. Think JSON or sqlite file. So that it won't depend on the browser cache

   ```bash
   git add .
   git commit -m "Fix button alignment in toolbar (#123)"
   git push -u origin fix/issue-123-button-alignment
   ```

6. **Open a Pull Request**:
   - Go to the original repository on GitHub
   - Click "New Pull Request"
   - Select your fork and branch
   - Describe what you changed and why
   - Reference the issue number if applicable

:::{tip} Using AI for Contributions
AI can help you understand existing code:

```
Explain how this widget's lifecycle works - especially the initialize()
and dispose() methods.
```

```
I want to add a new filter option to this panel. Show me where to add
the UI component and how to wire it to the existing filtering logic.
```

```
This extension uses React. Convert this example to use React best practices
while maintaining compatibility with the existing component structure.
```
:::

### 📋 Contribution checklist

Before opening a PR, verify:

- Code follows the project's style guide
- Changes are tested (automated tests and manual testing)
- Documentation is updated if needed
- Commit messages are clear and descriptive
- PR description explains what and why
- You've read and followed CONTRIBUTING.md

> - Tests pass in CI/CD (after opening PR)

Reviewer: No dead code! I see this often with LLM contributions
